### PR TITLE
Add ability to specify the go binary.

### PR DESCRIPTION
**Pull request by Bluebugs**
### Description:
Some go module only compile with specific version of go compiler. Being able to specify it is quite useful in this situation.

### Checklist:
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

#### Where applicable:
- [x] Public APIs match existing style and have Since: line.

### CODE REVIEW

**Review by andydotxyz — CHANGES_REQUESTED** [review]
I don't think `GOBIN` is sufficient is it?
Different versions of Go use different cache / package locations for their root so at least `GOROOT` is missing from this...

**Comment by Bluebugs** [comment]
In my experiment, it is the `go` command that read the GOROOT environment. So as long as the right binary is executed, it will pick the GOROOT environment variable. Is there any command line argument that you are thinking of? Because otherwise if they are environment variable, they will be propagated.

**Comment by andydotxyz** [comment]
I share the concern because, in my experience, just changing the binary path is often not sufficient

**Comment by Bluebugs** [comment]
Yes, but those are environment variable, so you set them when calling fyne in the same way you set GOBIN. There is nothing that the fyne command need to do about those, except to just let them pass through, which it does as default behavior.

**Comment by andydotxyz** [comment]
If I call a different `go` binary without setting `GOBIN` then I get a pile of compile errors with package versions not matching the compiler. and/or missing packages if the version difference is crossing something like `io/fs` etc.

**Comment by andydotxyz** [comment]
Oh, maybe that was a special case for when working on Go bugs.
Perhaps I am conflating my work in that with the more developer focused `PATH` vs `GOBIN`, will go read more.

**Comment by Bluebugs** [comment]
Do you mean, that you are changing the `go` binary itself? I am guessing that in this case you might have not properly updated the environment that those binary use. You can check it with `go env`.

I personally have multiple go installed, but they have different environment set for them (different GOROOT in my case).

**Comment by andydotxyz** [comment]
I think we are not talking about the same thing here.

From Go docs "Executables are installed in the directory named by the GOBIN environment variable, which defaults to $GOPATH/bin or $HOME/go/bin if the GOPATH environment variable is not set. "

So `GOBIN` refers to where user compiled Go executables should be installed... not the Go compiler.
https://pkg.go.dev/cmd/go

**Comment by Bluebugs** [comment]
Oh, indeed, I had missed that. Let me check.

**Comment by Bluebugs** [comment]
I picked an completely unused environment variable, GOEXEC. Is there any place where we are documenting the environment variable that fyne is picking up?

**Comment by andydotxyz** [comment]
Oh I see, you were looking for a new way to specify the compiler... Why not go with one of the ways that already work like GOROOT or suggesting an alias in the documentation.

Surely the best thing is if this picks up the Go version that is currently specified in the users environment?

**Comment by Bluebugs** [comment]
> Surely the best thing is if this picks up the Go version that is currently specified in the users environment?

Some OS provide multiple install of go in parallel. They do so by providing binary with different name (go1.20, go1.19, ...) and have them all in /usr/bin. Each binary pick its own GOROOT properly. The change is done on the binary name, not on the environment variable.

**Comment by andydotxyz** [comment]
> > Surely the best thing is if this picks up the Go version that is currently specified in the users environment?
> 
> Some OS provide multiple install of go in parallel. They do so by providing binary with different name (go1.20, go1.19, ...) and have them all in /usr/bin. Each binary pick its own GOROOT properly. The change is done on the binary name, not on the environment variable.

Some might, others don't. So it can be done by a) name change, b) PATH change and c) GOROOT setup.
I don't think we should design a new solution that works for some and not for others.
Why can't we avoid needing to support all of these and just pick up what Go the current env is using?

**Comment by Bluebugs** [comment]
On Linux, executable are installed in the /usr/bin from distribution package. 
- Changing their name require to manually modify the system after the installation or update of a distribution package.
- All the package get their binary installed in /usr/bin and so the executable have a different name otherwise they would override each other.
- GOROOT only affect the library used by the compiler, not the compiler itself. If you need to pick a different compiler, changing GOROOT does not have any effect on it.

I do not think that any of this solution are comparable with what this PR does. Could you clarify in which scenario this solution will not work?

**Comment by andydotxyz** [comment]
Sorry, I don't think I explained. I was not saying that a/b/c were ways to *set* the go version - I was saying that these can all be done by different installers and we need to /handle/ these (not require them).
Does that make more sense?

**Review by andydotxyz — APPROVED** [review]

File: cmd/fyne/internal/commands/build.go [hunk 1]
```diff
@@ -191,10 +192,17 @@ func (b *Builder) build() error {
 	fyneGoModRunner := b.runner
 	if b.runner == nil {
 		fyneGoModRunner = newCommand("go")
-		if goos != "gopherjs" {
-			b.runner = newCommand("go")
+		goBin := os.Getenv("GOBIN")
+		if goBin != "" {
+			log.Println("Using GOBIN environment variable for go command:", goBin)
+			fyneGoModRunner = newCommand(goBin)
```
[the remark this answers is not in the document]
Comment (Bluebugs) [reply cmd/fyne/internal/commands/build.go]: Seems like GitHub had some stale data on your side. The patch is using the `GO` environment variable as discussed before.

File: cmd/fyne/internal/commands/build.go [hunk 2]
```diff
@@ -191,10 +192,17 @@ func (b *Builder) build() error {
 	fyneGoModRunner := b.runner
 	if b.runner == nil {
 		fyneGoModRunner = newCommand("go")
-		if goos != "gopherjs" {
-			b.runner = newCommand("go")
+		goBin := os.Getenv("GOBIN")
+		if goBin != "" {
+			log.Println("Using GOBIN environment variable for go command:", goBin)
```
[the remark this answers is not in the document]
Comment (Bluebugs) [reply cmd/fyne/internal/commands/build.go]: That was removed. Very weird that you didn't get the last version of the patch.

File: cmd/fyne/internal/commands/build.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andydotxyz) [reply cmd/fyne/internal/commands/build.go]: Very strange indeed. The approval stands :)

File: cmd/fyne/internal/commands/build.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bluebugs) [reply cmd/fyne/internal/commands/build.go]: Oh, I got confused by GitHub, sorry for the very confused comment :-)

**Pull request merged**